core/src/impl/Kokkos_SICMSpace.cpp

```diff
@@ -0,0 +1,343 @@
+//              Copyright (2014) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <algorithm>
+#include <Kokkos_Macros.hpp>
+#if defined(KOKKOS_ENABLE_PROFILING)
+#include <impl/Kokkos_Profiling_Interface.hpp>
+#endif
+
+#include <cstddef>
+#include <cstdlib>
+#include <cstdint>
+#include <cstring>
+
+#include <iostream>
+#include <sstream>
+#include <cstring>
+
+#include <Kokkos_SICMSpace.hpp>
+#include <impl/Kokkos_Error.hpp>
+#include <Kokkos_Atomic.hpp>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Experimental {
+
+/* Default allocation mechanism */
+SICMSpace::SICMSpace()
+  : arena(nullptr)
+{
+  sicm_init();
+  arena = std::make_shared<sicm_arena>(static_cast<void *>(ARENA_DEFAULT));
+}
+
+SICMSpace::SICMSpace( sicm_device_list * devices )
+  : SICMSpace()
+{
+  if ( devices ) {
+    arena = std::shared_ptr <sicm_arena> (
+      new sicm_arena(sicm_arena_create(0, static_cast<sicm_arena_flags>(0), devices)),
+      []( sicm_arena * a ) { sicm_arena_destroy(*a); delete a; }
+    );
+  }
+
+  if ( *arena == ARENA_DEFAULT ) {
```

Comment:
`arena` is an `std::shared_ptr`, and needs to be dereferenced.

Reply:
@calccrypto said:
> arena is an std::shared_ptr, and needs to be dereferenced.

You are right, I totally missed that.

On that one though I am lucky because `ARENA_DEFAULT` is `NULL`, which implicit converts to `std::nullptr_t` so it happens to be comparable to any `shared_ptr` :)

You can ignore these comment on comparing to `ARENA_DEFAULT`.

Reply:
https://github.com/kokkos/kokkos/pull/2678#discussion_r377344722

Reply:
So right it should be `if ( arena.get() == ARENA_DEFAULT )` with the changes I suggested.